eth/fetcher/tx_fetcher.go

```diff
@@ -0,0 +1,291 @@
+		// There are many cases can lead to it:
+		// * We send the request to busy peer which can reply immediately
+		// * We send the request to malicious peer which doesn't reply deliberately
+		// * We send the request to normal peer for a batch of transaction, but some
+		//   transactions have been included into blocks. According to EIP these txs
+		//   won't be included.
+		// But it's fine to delete the fetching record and reschedule fetching iff we
+		// receive the annoucement again.
+		for hash, announce := range f.fetching {
+			if time.Since(announce.time) > txFetchTimeout {
+				delete(f.fetching, hash)
+				txFetchTimeoutMeter.Mark(1)
+			}
+		}
+		select {
+		case anno := <-f.notify:
+			txAnnounceInMeter.Mark(1)
+
+			// Drop the new announce if there are too many accumulated.
+			count := f.announces[anno.origin] + 1
+			if count > txAnnounceLimit {
+				txAnnounceDOSMeter.Mark(1)
+				break
+			}
+			// All is well, schedule the annouce if transaction is not yet downloading
+			if _, ok := f.fetching[anno.hash]; ok {
+				break
+			}
+			f.announces[anno.origin] = count
+			f.announced[anno.hash] = append(f.announced[anno.hash], anno)
+
+			if len(f.announced) == 1 {
+				f.reschedule(fetchTimer)
+			}
+			if f.announceHook != nil {
+				f.announceHook(anno.hash)
+			}
+		case <-fetchTimer.C:
+			// At least one tx's timer ran out, check for needing retrieval
+			request := make(map[string][]common.Hash)
+
+			for hash, announces := range f.announced {
+				if time.Since(announces[0].time) > arriveTimeout-gatherSlack {
+					// Pick a random peer to retrieve from, reset all others
+					announce := announces[rand.Intn(len(announces))]
+					f.forgetHash(hash)
+
+					// Skip fetching if we already receive the transaction.
+					if f.hasTx(hash) {
+						txAnnounceSkipMeter.Mark(1)
+						continue
+					}
+					// If the transaction still didn't arrive, queue for fetching
+					request[announce.origin] = append(request[announce.origin], hash)
+					f.fetching[hash] = announce
+				}
+			}
+			// Send out all block header requests
+			for peer, hashes := range request {
+				log.Trace("Fetching scheduled transactions", "peer", peer, "txs", hashes)
+
+				// Create a closure of the fetch and schedule in on a new thread
+				fetchTxs, hashes := f.fetching[hashes[0]].fetchTxs, hashes
+
+				go func() {
+					for {
+						// Split the hashes to several chunks if there are too many.
+						if len(hashes) > MaxTransactionFetch {
+							fetchTxs(hashes[:MaxTransactionFetch])
+							hashes = hashes[MaxTransactionFetch:]
+							txFetchOutMeter.Mark(int64(MaxTransactionFetch))
+						} else {
+							fetchTxs(hashes)
+							txFetchOutMeter.Mark(int64(len(hashes)))
+							return
+						}
+					}
+				}()
+			}
+			// Schedule the next fetch if blocks are still pending
+			f.reschedule(fetchTimer)
+		case op := <-f.inject:
+			// The transaction from all eth protocol peers will arrive here.
+			var drop bool
+			errs := f.addTxs(op.txs)
```

Comment:
In the original code txs were added on the peer's handler thread. This allowed them to be deduplicated and recovered concurrently. Moving addTxs here serializes these ops onto the loop thread, which is probably something that we don't want. Could we call addTxs in `EnqueueTxs` and only "inject" the hash + error for cleanup? We can also rename "inject" to "cleanup" perhaps, since in that case this code path only removes internal indices.